### ISSUE TITLE
OpenSSL 3.2.0-alpha1: argon2 impl uses C11 _Static_assert

### ISSUE THREAD

**Issue by rainerjung** [openssl/openssl#22017]
When compiling on RHEL 6 or SLES 11 I get the following error:

```
./libcrypto.a(libdefault-lib-argon2.o): In function `kdf_argon2_ctx_set_out_length':
.../providers/implementations/kdfs/argon2.c:1225: undefined reference to `_Static_assert'
./libcrypto.a(libdefault-lib-argon2.o): In function `kdf_argon2_ctx_set_t_cost':
.../providers/implementations/kdfs/argon2.c:1189: undefined reference to `_Static_assert'
./libcrypto.a(libdefault-lib-argon2.o): In function `kdf_argon2_ctx_set_m_cost':
.../providers/implementations/kdfs/argon2.c:1204: undefined reference to `_Static_assert'
```

These platforms provide GCC 4.3 resp. 4.4 which does not support _Static_assert (seems to have been introduced by C11).

The file actually uses ossl_static_assert_type_eq, which is defined in include/internal/e_os.h as:

```
/* ossl_static_assert_type_eq: gcc-only variable type static assertion */
# if defined(__GNUC__) && !defined(__clang__)
#  define ossl_static_assert_type_eq(type, x)                                \
        _Static_assert((__builtin_types_compatible_p(type, __typeof__(x))),  \
                        #x " type check failed, expected: " #type)
# else
#  define ossl_static_assert_type_eq(type, x)
# endif
```

So there seems to be no check for availability of _Static_assert.

**Comment by paulidale** [comment]
Does the change in #22022 fix your problem?

**Comment by mattcaswell** [comment]
My proposed fix for this in #22091
